src/dstack/_internal/server/services/logs.py

```diff
@@ -169,24 +180,70 @@ def write_logs(
             )
 
     def _write_logs(self, stream: str, log_events: List[RunnerLogEvent]) -> None:
-        events = [self._runner_log_event_to_cloudwatch_event(event) for event in log_events]
-        params = {
-            "logGroupName": self._group,
-            "logStreamName": stream,
-            "logEvents": events,
-        }
         with self._wrap_boto_errors():
             self._ensure_stream_exists(stream)
             try:
-                self._client.put_log_events(**params)
+                self._put_log_events(stream, log_events)
                 return
             except botocore.exceptions.ClientError as e:
                 if not self._is_resource_not_found_exception(e):
                     raise
                 logger.debug("Stream %s not found, recreating", stream)
             # The stream is probably deleted due to retention policy, our cache is stale.
             self._ensure_stream_exists(stream, force=True)
-            self._client.put_log_events(**params)
+            self._put_log_events(stream, log_events)
+
+    def _put_log_events(self, stream: str, log_events: List[RunnerLogEvent]) -> None:
+        for batch in self._get_batch_iter(stream, log_events):
+            self._client.put_log_events(
+                logGroupName=self._group,
+                logStreamName=stream,
+                logEvents=batch,
+            )
+
+    def _get_batch_iter(
+        self, stream: str, log_events: List[RunnerLogEvent]
+    ) -> Iterator[List[_CloudWatchLogEvent]]:
+        shared_event_iter = iter(log_events)
+        event_iter = shared_event_iter
+        while True:
+            batch, excessive_event = self._get_next_batch(stream, event_iter)
+            if not batch:
+                return
+            yield batch
+            if excessive_event is not None:
+                event_iter = itertools.chain([excessive_event], shared_event_iter)
+            else:
+                event_iter = shared_event_iter
+
+    def _get_next_batch(
+        self, stream: str, event_iter: Iterator[RunnerLogEvent]
+    ) -> Tuple[List[_CloudWatchLogEvent], Optional[RunnerLogEvent]]:
+        batch: List[_CloudWatchLogEvent] = []
+        total_size = 0
+        event_count = 0
+        for event in event_iter:
+            # Normally there should not be empty messages.
+            if not event.message:
+                continue
+            cw_event = self._runner_log_event_to_cloudwatch_event(event)
+            # as message is base64-encoded, length in bytes = length in code points.
+            message_size = len(cw_event["message"]) + self.MESSAGE_OVERHEAD_SIZE
+            if message_size > self.MESSAGE_MAX_SIZE:
+                logger.warning(
+                    "Stream %s: skipping event %d, message exceeds max size",
+                    stream,
+                    event.timestamp,
+                )
+                continue
```

Comment:
If it ever happens, I think skipping the event will be quite unexpected. The user just won't find the message in the run logs and the server admin will have to deal with warnings they can't do anything about.

Maybe split such long messages into several events or cut them?

Reply:
It turned out that we should never hit this limit, as the underlying code in the runner uses 32KiB buffer when copying from PTY to log writer. I've increased log level to ERROR and added a comment explaining where this assumption comes from.